### ISSUE TITLE
Enabling PropagateDeps on the workload leave behind RB if no clusters are available for scheduling and then delete workload

### ISSUE THREAD

**Issue by XiShanYongYe-Chang** [karmada-io/karmada#5942]
**What happened**:

When Workload has PropagateDeps enabled and there are no suitable clusters for scheduling, deleting the Workload may result in leftover ResourceBinding.

**What you expected to happen**:

ResourceBinding is cleaned up properly.

**How to reproduce it (as minimally and precisely as possible)**:

1. Deploy the following resources in Karmada

<details>
<summary>deploy.yaml</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deploy-dep
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx:1.19.0
        imagePullPolicy: IfNotPresent
        name: nginx
      volumes:
      - configMap:
          defaultMode: 420
          name: configmap-zrnpn
        name: vol-configmap
```

</details>

<details>
<summary>cm.yaml</summary>

```yaml
apiVersion: v1
data:
  user: karmada
kind: ConfigMap
metadata:
  name: configmap-zrnpn
```
</details>

<details>
<summary>pp.yaml</summary>

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: deploy-dep
spec:
  conflictResolution: Abort
  placement:
    clusterAffinity:
      clusterNames:
      - member1
  preemption: Never
  priority: 0
  propagateDeps: true
  resourceSelectors:
  - apiVersion: apps/v1
    kind: Deployment
    name: deploy-dep
  schedulerName: default-scheduler
```
</details>

2. Modify the PropagationPolicy resource to set the cluster to a non-existent cluster

<details>
<summary>pp.yaml</summary>

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: deploy-dep
spec:
  conflictResolution: Abort
  placement:
    clusterAffinity:
      clusterNames:
      - member-no-exist
  preemption: Never
  priority: 0
  propagateDeps: true
  resourceSelectors:
  - apiVersion: apps/v1
    kind: Deployment
    name: deploy-dep
  schedulerName: default-scheduler
```
</details>

3. Delete the Deployment resource

And then you can find that the related ResourceBinding resource was left behind.

<details>
<summary>rb.yaml</summary>

```yaml
apiVersion: work.karmada.io/v1alpha2
kind: ResourceBinding
metadata:
  annotations:
    policy.karmada.io/applied-placement: '{"clusterAffinity":{"clusterNames":["member-no-exist"]},"clusterTolerations":[{"key":"cluster.karmada.io/not-ready","operator":"Exists","effect":"NoExecute","tolerationSeconds":30},{"key":"cluster.karmada.io/unreachable","operator":"Exists","effect":"NoExecute","tolerationSeconds":30}]}'
    propagationpolicy.karmada.io/name: deploy-dep
    propagationpolicy.karmada.io/namespace: default
    resourcebinding.karmada.io/dependencies: '[{"apiVersion":"v1","kind":"ConfigMap","namespace":"default","name":"configmap-zrnpn"}]'
  creationTimestamp: "2024-12-12T03:37:59Z"
  deletionGracePeriodSeconds: 0
  deletionTimestamp: "2024-12-12T03:42:08Z"
  finalizers:
  - karmada.io/binding-dependencies-distributor
  generation: 6
  labels:
    propagationpolicy.karmada.io/permanent-id: 607f4ef1-934b-425c-bc7b-334f8116dda3
    resourcebinding.karmada.io/permanent-id: 7837ddcc-58ec-4ba4-83d5-6aa459551375
  name: deploy-dep-deployment
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Deployment
    name: deploy-dep
    uid: cff09982-1eca-44b5-aae5-ef97150b8b32
  resourceVersion: "200733"
  uid: ada763f4-9a36-45e8-ac0a-2b044640b726
spec:
  conflictResolution: Abort
  placement:
    clusterAffinity:
      clusterNames:
      - member-no-exist
    clusterTolerations:
    - effect: NoExecute
      key: cluster.karmada.io/not-ready
      operator: Exists
      tolerationSeconds: 30
    - effect: NoExecute
      key: cluster.karmada.io/unreachable
      operator: Exists
      tolerationSeconds: 30
  propagateDeps: true
  replicas: 3
  resource:
    apiVersion: apps/v1
    kind: Deployment
    name: deploy-dep
    namespace: default
    resourceVersion: "200005"
    uid: cff09982-1eca-44b5-aae5-ef97150b8b32
  schedulerName: default-scheduler
status:
  conditions:
  - lastTransitionTime: "2024-12-12T03:40:50Z"
    message: '0/3 clusters are available: 3 cluster(s) did not match the placement
      cluster affinity constraint.'
    reason: NoClusterFit
    status: "False"
    type: Scheduled
  - lastTransitionTime: "2024-12-12T03:40:50Z"
    message: Failed to apply all works, see status.aggregatedStatus for details
    reason: FullyAppliedFailed
    status: "False"
    type: FullyApplied
  lastScheduledTime: "2024-12-12T03:37:59Z"
  schedulerObservedGeneration: 3
```
</details>

The direct reason why resourceBinding remains is because it still has a finalizer `karmada.io/binding-dependencies-distributor` that has not been cleaned up.

**Anything else we need to know?**:

The finalizer `karmada.io/binding-dependencies-distributor`  is added and removed by `Dependencies Distributor`:

https://github.com/karmada-io/karmada/blob/471d850a1985acf6ee79f7d21b3dee5b33466204/pkg/dependenciesdistributor/dependencies_distributor.go#L286-L289

https://github.com/karmada-io/karmada/blob/471d850a1985acf6ee79f7d21b3dee5b33466204/pkg/dependenciesdistributor/dependencies_distributor.go#L258-L266

From the analysis of the phenomenon and logs, when ResourceBinding is marked for deletion (DeletionTimestamp is marked with timestamp), it does not enter the Reconcile processing flow of `Dependencies Distributor`.

In addition, the DeletionTimestamp of the resource is marked as an update event, not a deletion event, so it can be concluded that the update event has been filtered out.

https://github.com/karmada-io/karmada/blob/471d850a1985acf6ee79f7d21b3dee5b33466204/pkg/dependenciesdistributor/dependencies_distributor.go#L651-L664

Looking at the filtering logic of update, it is filtered by this logic:

https://github.com/karmada-io/karmada/blob/471d850a1985acf6ee79f7d21b3dee5b33466204/pkg/dependenciesdistributor/dependencies_distributor.go#L659-L663

**Environment**:
- Karmada version: master branch(before v1.12, include v1.12)
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

**Comment by XiShanYongYe-Chang** [comment]
I currently think of two ideas. 

The first idea is for update events, if the DeletionTimestamp of the resource is not empty, it means that it is in the deletion process and will not be filtered.

The second idea is to remove the following filter:

https://github.com/karmada-io/karmada/blob/master/pkg/dependenciesdistributor/dependencies_distributor.go#L659-L663

This filtering logic does not match the content described in the annotation, and it is not equivalent to the resource not being scheduled.

Removing this logic will add some reconcile, but the number of increases is limited, so I understand that the impact on performance will not be huge

**Comment by XiShanYongYe-Chang** [comment]
> note: When the DeletionTimestamp of the cr resource is set, the generation of the resource will be increased by 1.

**Comment by RainbowMango** [comment]
I prefer the second approach, which means removing [the filter](https://github.com/karmada-io/karmada/blob/master/pkg/dependenciesdistributor/dependencies_distributor.go#L659-L663). 

Since we introduced the finalizer, we can't filter events based on whether the RB has been scheduled. When deleting RB it might not have been scheduled(but has been reconciled at least once, and the finalizer has been added).
